vercel/turborepo
# feat(tui): persistent preferences

### Description

Users have expressed that they find themselves repeatedly having to select out the task that they want to use. This ends up being annoying when invoking `turbo`, shutting down, invoking `turbo`, shutting down...You end up having to reselect the task you care about every re-invoke.

To improve on this, we're going to hold some state in `.turbo/preferences/tui.json` so that users pick up where they left off from their previous TUI invocation. State we'll keep around is:
- The selected task name (Responsible for the "pinned-ness" portion of state)
- Toggled state of task list visibility

### Testing Instructions

Wrote tests but you should also try it out by hand. You should be able to run `turbo` invocations, pick out tasks, and see the state persist across invocations. You also should be able to delete the `.turbo/preferences/tui.json` file and it'll feel like the first invocation (no preferences).

### Small note
One thing that I noticed while hand-testing is that this implementation tolerates extra fields fine, but doesn't tolerate the wrong types for a field. It will crash for in the latter scenario.

Personally, I think this is fine. You'd have to purposefully go and edit the file to get the wrong types into there.
